components/expand-collapse/expand-collapse.js

```diff
@@ -0,0 +1,140 @@
+import { css, html, LitElement } from 'lit-element/lit-element.js';
+import ResizeObserver from 'resize-observer-polyfill/dist/ResizeObserver.es.js';
+import { styleMap } from 'lit-html/directives/style-map.js';
+
+const reduceMotion = matchMedia('(prefers-reduced-motion: reduce)').matches;
+
+const states = {
+	COLLAPSING: 'collapsing', // in the process of collapsing
+	COLLAPSED: 'collapsed', // fully collapsed
+	EXPANDING: 'expanding', // in the process of expanding
+	EXPANDED: 'expanded', // fully expanded
+};
+
+class ExpandCollapse extends LitElement {
+
+	static get properties() {
+		return {
+			expanded: { type: Boolean, reflect: true },
+			_height: { type: Number },
+			_state: { type: String }
+		};
+	}
+
+	static get styles() {
+		return css`
+			:host {
+				display: block;
+			}
+
+			:host([hidden]) {
+				display: none;
+			}
+
+			.d2l-expand-collapse-container {
+				display: none;
+				height: 0px;
+				overflow: hidden;
+				transition: height 400ms cubic-bezier(0, 0.7, 0.5, 1);
+			}
+
+			.d2l-expand-collapse-container:not([data-state="collapsed"]) {
+				display: block;
+			}
+
+			.d2l-expand-collapse-container[data-state="expanded"] {
+				overflow: visible;
+			}
+
+			/* prevent margin colapse on slotted children */
+			.d2l-expand-collapse-content:before,
+			.d2l-expand-collapse-content:after {
+				content: ' ';
+				display: table;
+			}
+
+			@media (prefers-reduced-motion: reduce) {
+				.d2l-expand-collapse-container {
+					transition: none;
+				}
+			}
+		`;
+	}
+
+	constructor() {
+		super();
+		this._onContentResize = this._onContentResize.bind(this);
+		this.expanded = false;
+		this._state = states.COLLAPSED;
+	}
+
+	get expanded() {
+		return this._expanded;
+	}
+
+	set expanded(val) {
+		const oldVal = this._expanded;
+		if (oldVal !== val) {
+			this._expanded = val;
+			this.requestUpdate('expanded', oldVal);
+			this._expandedChanged(val);
+		}
+	}
+
+	firstUpdated() {
+		super.firstUpdated();
+
+		const content = this.shadowRoot.querySelector('.d2l-expand-collapse-content');
+		this._resizeObserver = new ResizeObserver(this._onContentResize);
+		this._resizeObserver.observe(content);
+	}
+
+	render() {
+		const styles = {
+			height: this.expanded ? `${this._height}px` : null
+		};
+		return html`
+			<div class="d2l-expand-collapse-container" data-state=${this._state} @transitionend=${this._onTransitionEnd} style=${styleMap(styles)}>
+				<div class="d2l-expand-collapse-content">
+					<slot></slot>
+				</div>
+			</div>
+		`;
+	}
+
+	_expandedChanged(val) {
+		if (val) {
+			this._state = reduceMotion ? states.EXPANDED : states.EXPANDING;
+			this._updateHeight();
+		} else {
+			this._state = reduceMotion ? states.COLLAPSED : states.COLLAPSING;
+			this._height = null;
+		}
```

Comment:
I feel like this component (or maybe it should be a mixin) would become really powerful if it could handle `aria-expanded` for us. The W3C [has an examplle](https://www.w3.org/WAI/GL/wiki/Using_aria-expanded_to_indicate_the_state_of_a_collapsible_element) of how this should work. But that would imply that the component/mixin is aware of the "opener" (the thing that when clicked toggles the state).

Ideally in my mind, the reusable stuff that'd be valuable for other components:
- The animation of the content area (which you have)
- Wire-up to an opener that toggles things and sets `aria-expanded` on it
- Ability for the opener to be anything and styled in any way

Basically... if we look at more/less and at the design for collapsible alerts and the design for the FACE panels... if we could refactor whatever they share into a mixin/component, what would it look like?

Reply:
I did explore making it a mixin but the big drawback there is that we either need to either manipulate the DOM directly by adding styles outside of render or create methods people would need to call themselves as part of render, both of which felt bad if people are going to be using this outside of core.

How would you feel about creating something like a `d2l-expand-collapse-opener` that behaves in basically the same manner as the `d2l-dropdown-opener`? Then that could be used along with this to accomplish all of those goals.

Reply:
Or I guess if we are following the `dropdown` model for adding openers it would be something along the lines of renaming this to `d2l-expand-collapse-content` then adding `d2l-expand-collapse` for the opener logic.

Reply:
Yeah it's complicated eh. Like the opener needs to be able to be literally anything... typically a button of some kind of with various stylings and sometimes the whole "bar" is the click target. But I don't believe we need to have `id` references... unless the `aria-controls` needs to be used.

Reply:
I think we'd need some sort of `id` reference just as some way to keep the opener and content in sync. At least it feels like it'd be the simplest option if we could just do something like `<d2l-expand-collapse opener="my-opener-id">` then it could find the opener element and add click handlers and set the `aria-expanded` attribute.

It wouldn't have the same shadow boundary ID issues that tooltip has since the actual linking would be done using Javascript.

Reply:
Was just thinking we'd do what dropdown does where it uses a CSS class applied to an element somewhere inside a slot.

Reply:
The reason I don't think that works in this case is because that also has a dependency on the `d2l-dropdown-content` being inside the same slot. That works for dropdowns because they are absolutely positioned so where the `d2l-dropdown-content` is in the DOM tree doesn't really matter.

With an expand collapse area that becomes an issue because there is no guarantee that the opener and the collapsible section are next to each other in the DOM:
![Capture](https://user-images.githubusercontent.com/11379933/81605890-24393200-93a0-11ea-9e41-deefb93662e4.PNG)

Reply:
Yeah true! So in that case, some 3rd party is going to need to "marshall" the click event from the opener/closer (which may be 2 different elements like in this example) and notify the expand/collapse container. It _could_ pass the opener to the container as a DOM node and the container could apply some ARIA to it... but at that point we're getting diminishing returns.

Similarly, in theory the container could expand or collapse based on another party interacting with it, and that event would need to be marshalled back to the opener/closer.

So essentially I think I've come around to the container itself being the interesting bit of reusable UI here I think!

Reply:
👀 ... 🤔